internal/engine/local/probe.go

```diff
@@ -0,0 +1,125 @@
+	TCPSocket(host string, port int) prober.ProberFunc
+	Exec(name string, args ...string) prober.ProberFunc
+}
+
+func probeWorkerFromSpec(manager ProberManager, probeSpec *v1.Probe, opts ...probe.WorkerOption) (*probe.Worker, error) {
+	probeFunc, err := proberFromSpec(manager, probeSpec)
+	if err != nil {
+		return nil, err
+	}
+
+	if probeSpec.InitialDelaySeconds >= 0 {
+		opts = append(opts, probe.WorkerInitialDelay(time.Duration(probeSpec.InitialDelaySeconds)*time.Second))
+	}
+	if probeSpec.TimeoutSeconds > 0 {
+		opts = append(opts, probe.WorkerTimeout(time.Duration(probeSpec.TimeoutSeconds)*time.Second))
+	}
+	if probeSpec.PeriodSeconds > 0 {
+		opts = append(opts, probe.WorkerPeriod(time.Duration(probeSpec.PeriodSeconds)*time.Second))
+	}
+
+	if probeSpec.SuccessThreshold > 0 {
+		opts = append(opts, probe.WorkerSuccessThreshold(int(probeSpec.SuccessThreshold)))
+	}
+	if probeSpec.FailureThreshold > 0 {
+		opts = append(opts, probe.WorkerFailureThreshold(int(probeSpec.FailureThreshold)))
+	}
+
+	w := probe.NewWorker(probeFunc, opts...)
+	return w, nil
+}
+
+func proberFromSpec(manager ProberManager, probeSpec *v1.Probe) (prober.Prober, error) {
+	if probeSpec == nil {
+		return nil, nil
+	} else if probeSpec.Exec != nil {
+		return manager.Exec(probeSpec.Exec.Command[0], probeSpec.Exec.Command[1:]...), nil
+	} else if probeSpec.HTTPGet != nil {
+		u, err := extractURL(probeSpec.HTTPGet)
+		if err != nil {
+			return nil, err
+		}
+		return manager.HTTPGet(u, convertHeaders(probeSpec.HTTPGet.HTTPHeaders)), nil
+	} else if probeSpec.TCPSocket != nil {
+		port, err := extractPort(probeSpec.TCPSocket.Port)
+		if err != nil {
+			return nil, err
+		}
+		return manager.TCPSocket(probeSpec.TCPSocket.Host, port), nil
+	}
+
+	return nil, ErrUnsupportedProbeType
+}
+
+// extractURL converts a K8s HTTP GET probe spec to a Go URL
+// adapted from https://github.com/kubernetes/kubernetes/blob/v1.20.2/pkg/kubelet/prober/prober.go#L163-L186
+func extractURL(httpGet *v1.HTTPGetAction) (*url.URL, error) {
+	port, err := extractPort(httpGet.Port)
+	if err != nil {
+		return nil, err
+	}
+	u, err := url.Parse(httpGet.Path)
+	if err != nil {
+		return nil, err
+	}
+	u.Scheme = strings.ToLower(string(httpGet.Scheme))
+	if u.Scheme == "" {
+		u.Scheme = "http"
+	}
+	if httpGet.Host == "" {
+		return nil, errors.New("no host specified")
+	}
+	u.Host = net.JoinHostPort(httpGet.Host, strconv.Itoa(port))
+	return u, nil
+}
+
+// extractPort converts a K8s multi-type value to a valid port number or returns an error.
+// adapted from https://github.com/kubernetes/kubernetes/blob/v1.20.2/pkg/kubelet/prober/prober.go#L203-L223
+// (note: this implementation is substantially simplified from K8s - it does not handle "named" ports nor implicit
+//  string conversion, as these are not relevant concerns here)
+func extractPort(v intstr.IntOrString) (int, error) {
+	var port int
+	if v.Type == intstr.Int {
+		port = v.IntValue()
+	}
```

Comment:
If you pass `foo`, you'll actually get an error of `invalid port number: foo` -- the error message uses `instsr::String()` which returns the raw string value or an itoa'd version of the int. I'll add some comments to this function - can also rewrite to be more explicit. (There is a test case for the string case + error message which should hopefully catch any regressions on that front as well)

Reply:
Let me know if the tweaks in a4128449c4238f1a318fd3e04f1ff673773de29c are clear enough

Reply:
if we pass "1234", do we get "invalid port number: 1234"?

Reply:
Yeah, you would, which is icky. K8s actually has a fallback to the fallback that'll attempt to `atoi` - I'll do the same for consistency. (FYI you actually _won't_ ever have a string at this point currently because the port is an int in the Starlark grammar so it'll reject strings before ever creating a `*v1.Probe`.)

Reply:
Okay cleaned this up in 5673041e2c75594260a32579380400d76baf3ff3 - you'll get `invalid port number: "http"` for random strings and `port number out of range: 65536` for any valid ints (or stringified ints) that are not valid ports.